teams_rules.go

```diff
@@ -0,0 +1,222 @@
+	YTRestricted TeamsGatewayAction = "ytrestricted"
+	On           TeamsGatewayAction = "on"
+	Off          TeamsGatewayAction = "off"
+	Scan         TeamsGatewayAction = "scan"
+	NoScan       TeamsGatewayAction = "noscan"
+	Isolate      TeamsGatewayAction = "isolate"
+	NoIsolate    TeamsGatewayAction = "noisolate"
+	Override     TeamsGatewayAction = "override"
+	L4Override   TeamsGatewayAction = "l4_override"
+)
+
+// TeamsRule represents an Teams wirefilter rule.
+type TeamsRule struct {
+	ID          string             `json:"id,omitempty"`
+	CreatedAt   *time.Time         `json:"created_at,omitempty"`
+	UpdatedAt   *time.Time         `json:"updated_at,omitempty"`
+	DeletedAt   *time.Time         `json:"deleted_at,omitempty"`
+	Name        string             `json:"name"`
+	Description string             `json:"description"`
+	Precedence  uint64             `json:"precedence"`
+	Enabled     bool               `json:"enabled"`
+	Action      TeamsGatewayAction `json:"action"`
+	Filters     []TeamsFilterType  `json:"filters"`
+	Traffic     string             `json:"traffic"`
+	Identity    string             `json:"identity"`
+	Version     uint64             `json:"version"`
+	Settings    TeamsRuleSettings  `json:"rule_settings,omitempty"`
+}
+
+// TeamsRuleResponse is the API response, containing a single rule
+type TeamsRuleResponse struct {
+	Response
+	Result TeamsRule `json:"result"`
+}
+
+// TeamsRuleResponse is the API response, containing an array of rules
+type TeamsRulesResponse struct {
+	Response
+	Result []TeamsRule `json:"result"`
+}
+
+// TeamsRulePatchRequest is used to patch an existing rule
+type TeamsRulePatchRequest struct {
+	UUID        string             `json:"id"`
+	Name        string             `json:"name"`
+	Description string             `json:"description"`
+	Precedence  uint64             `json:"precedence"`
+	Enabled     bool               `json:"enabled"`
+	Action      TeamsGatewayAction `json:"action"`
+	Settings    TeamsRuleSettings  `json:"rule_settings,omitempty"`
+}
+
+// TeamsRules returns all rules within an account.
+//
+// API reference: https://api.cloudflare.com/#teams-rules-properties
+func (api *API) TeamsRules(ctx context.Context, accountID string) ([]TeamsRule, error) {
+	uri := fmt.Sprintf("/accounts/%s/gateway/rules", accountID)
+
+	res, err := api.makeRequestContext(ctx, http.MethodGet, uri, nil)
+	if err != nil {
+		return []TeamsRule{}, err
+	}
+
+	var teamsRulesResponse TeamsRulesResponse
+	err = json.Unmarshal(res, &teamsRulesResponse)
+	if err != nil {
+		return []TeamsRule{}, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return teamsRulesResponse.Result, nil
+}
+
+// TeamsRule returns the rule with rule id in the url (uuid without dashes)
+//
+// API reference: https://api.cloudflare.com/#teams-rules-properties
+func (api *API) TeamsRule(ctx context.Context, accountID string, ruleId string, ) (TeamsRule, error) {
+	uri := fmt.Sprintf("/accounts/%s/gateway/rules/%s", accountID, ruleId)
```

Comment:
looking between the API and this PR, we've got a difference in how the rule ID is stored and retrieved.

in the API (the RESTful route), we want a hash style value, however on the resource itself it is a UUIDv4 value. this is very confusing and will make handling this in Terraform/cf-terraforming very painful as the identifier is used as an anchor for subsequent API calls. ideally, we decide on one and continue with it to avoid extra work going forward.

Reply:
Turns we have new change in api, that accepts uuid with dashes. I have updated the comment. Resolving this.